bookwyrm/views/status.py

```diff
@@ -115,6 +115,18 @@ def post(self, request, status_type, existing_status_id=None):
         if status.reply_parent:
             status.mention_users.add(status.reply_parent.user)
 
+        # inspect the text for hashtags
+        for (mention_text, mention_hashtag) in find_hashtags(content).items():
+            # add them to status mentions fk
+            status.mention_hashtags.add(mention_hashtag)
+
+            # turn the mention into a link
+            content = re.sub(
```

Comment:
This is what I wanted to test with my local federation setup.

By parsing and linking to the hashtag on status creation, this means that when I create a status with a hashtag on instance A, when this tag is federated to instance B the hashtag will link back to instance A.

Not sure if I like this, I think I would much rather have hashtags be linked to a local hashtag page. Which would probably require some rethinking of how and when this is happening. Curious what you think about this @mouse-reeve?

Reply:
I also think it would be preferable to stay on your instance when you click a hashtag. The base model has a [property that replaces the domain in a path with the local path](https://github.com/bookwyrm-social/bookwyrm/blob/main/bookwyrm/models/base_model.py#L54), which I think you could use by just replacing `.remote_id` with `.local_path`

Reply:
Oh, interesting! thank you for the pointer.
In that case, the best way forward seems to be to rework the hashtag URLs and calculate a unique slug to ensure that the path of the `.remote_id` stays identical across different instances since it currently includes `Hashtag.id`.

I'll get this in a mergeable shape eventually! 😄

Reply:
oh yes, there's a way to do that too, by overwriting [`get_remote_id`](https://github.com/bookwyrm-social/bookwyrm/blob/main/bookwyrm/models/base_model.py#L54) on the model. [`Shelf`](https://github.com/bookwyrm-social/bookwyrm/blob/main/bookwyrm/models/shelf.py#L65) is a model that uses this

Reply:
I have been thinking about this a bit more. Sorry that I don't come back to this PR that often, it's quite a big one already and requires loading a lot of context into my brain every time I start.

Please let me know if this sounds way off or if you see a better way forward.

The hashtag federation will work quite differently from the other first-party objects because there's no reason to federate hashtags on their own, but only as part of a federated status. Especially in comparison with, say, Shelf or similar models.
Wouldn't this mean that for each status that comes into a user's inbox we would have to extend the importing process with:
1. parsing the status content for hashtags, or hashtag links
    * I'm not sure if I already added hashtags to the status' ActivityPub representation as tags ([similar to how Mastodon does it](https://docs.joinmastodon.org/spec/activitypub/#status))
3. find existing or create new local hashtags
4. replace the URLs in the HTML-rendered status content with the new local hashtag URL

The only other object that I've seen traces of in the code but never managed to use myself is book mentions. The federation logic for those should be identical to hashtags. But I've never managed to use them, that being said, I haven't tried _that_ hard.

And to add to all of this complexity, this only affects the federation between two BookWyrm instances, seeing how hashtags might work when federating from BookWyrm to other ActivityPub services like Mastodon, or a lesser degree even a Mastodon status to BookWyrm.

There are quite a few things that still need thinking about, not sure if it would be easier in a synchronous conversation or pairing session, if you're up for it and if we can make it work timezone-wise.

Reply:
alright after quite a few hours stepping through the ActivityPub code in a debugger I _think_ I have a fairly good solution to this now, it's in 84ffe3ba2247b711bc28c138bdad1a4b1f3c1053. It still needs proper test coverage though, which I'm hoping to get to in the next few days.

Reply:
I don't think book mentions are used anywhere at this point? They used to be in the past. (I'm not sure if this if this is true but that's my suspicion).

I think your solution looks good! And while I didn't see your original comment in a timely manner (oops), I am super down to talk about things synchronously in general :)